hummingbird.py

```diff
@@ -0,0 +1,153 @@
+from os import remove
+from sys import stdin
+import datetime
+import random
+import string
+import ast
+import threading
+import requests
+import config
+from network import get_MAC, print_MAC_address
+from music import MusicPlayer
+from utils import log, generate_random_suffix, is_unknown_user
+
+
+class User:
+	def __init__(self, system, name=config.unknown_user_prefix, song=config.need_to_assign, length=config.time_max_song_length, arrival=datetime.datetime.now()):
+		if is_unknown_user(name=name):
+			self.name = name + generate_random_suffix()
+		else:
+			self.name = name
+		if song == config.need_to_assign:
+			self.song = system.music_player.get_random_song()
+		else:
+			self.song = song
+		self.length = length
+		self.arrival = arrival
+
+	def has_not_played_today(self):
+		# not played today if:
+		# 	1) it's now after reset point and last played earlier today before reset point or earlier
+		# 	2) it's now before reset point and last played earlier yesterday before reset point or earlier
+		now = datetime.datetime.now()
+		today = now.date()
+		yesterday = today - datetime.timedelta(days=1)
+		return (
+			now.time() >= config.time_reset_time and self.arrival < datetime.datetime.combine(today, config.time_reset_time)
+		) or (
+			now.time() < config.time_reset_time and self.arrival < datetime.datetime.combine(yesterday, config.time_reset_time)
+		)
+
+	def queue_song(self, music_player):
+		self.arrival = datetime.datetime.now()
+		music_player.queue_song(user=self)
+
+	def __unicode__(self):
+		return "{name}: {song} ({length}s)".format(name=self.name, song=self.song, length=self.length)
+
+class System:
+	def __init__(self):
+		log(message="Initializing...")
+		# local_payload_cache uses mac_id ("address") as key, with the value being a dictionary of mac_id and datetime value of last time the api was pinged.
+		# Cache is only used if use_cache is set to True in config.py
+		self.local_payload_cache = {}		
+		self.music_player = MusicPlayer()
+		self.all_addresses = self.read_in_addresses()
+		self.waiting_for_input = False
+		self.input_timeout()
+		log(message="Waiting for tcpdump to provide input...")
+
+		while True:
+			# handle input as network traffic from tcpdump
+			addresses = get_MAC(line=stdin.readline())
+			self.waiting_for_input = False
+			for address in addresses:
+				
+				## Pings Hummingbird Django server. If use_cache is set to True, use self.local_payload_cache as a cache.
+				if config.use_cache:
+					if address.lower() in self.local_payload_cache:
+						# payload is a tuple of address and last_checked_datetime
+						payload = self.local_payload_cache[address.lower()]
+						# Set a 60 second cache				
+						if (datetime.datetime.now() - payload['last_sent_dt']).seconds > config.cache_time_seconds:
+							r = requests.get("http://127.0.0.1:8000/hummingbird/build_user_from_device/", params=payload)
```

Comment:
Maybe we could move the request URL to settings so that it can be changed easily. It's something I could imagine users wanting to change if port 8000 is already taken.